### PR TITLE
Issue #324, load validation data with generator instead of storing it. 

### PR DESCRIPTION
**Pull request by Arkkienkeli**
Supposed to solve #324. 

I checked it for several crop generators, but it does not work for old crop generator (in fact, it does not work already) and full image crop generator also requires additional testing. `sampled_crop_generator` and `online_labels_crop_generator` seem to work.

### CODE REVIEW

**Comment by Arkkienkeli** [comment]
@jccaicedo 
I made an update we discussed and also did run for a few epochs in TA-ORF dataset (cells out strongest treatments set). The results below seem to be good to go.
Full image crop generator also seems to run (ran for 2 epochs in full combined set).

Original 
```
epoch,acc,average_class_precision,loss,top_5,val_acc,val_average_class_precision,val_loss,val_top_5
0,0.14688082,0.240007,3.5200090759071334,0.39571694,0.21708238,0.4543684,3.1999364354184285,0.510945
1,0.26178423,0.53904086,3.0034031226426308,0.57122904,0.27354497,0.5721288,2.9709691701879826,0.57792014
2,0.3148132,0.60038793,2.776936955569621,0.6301792,0.31678003,0.6305915,2.7624577151404486,0.6391046
3,0.35271183,0.6425584,2.6179563534104404,0.67138034,0.31254724,0.65187323,2.8343186240265337,0.61945736
4,0.38531482,0.65860534,2.4845399945363873,0.7036342,0.35668933,0.66472155,2.6330002909121304,0.6754
```
After update
```
epoch,acc,average_class_precision,loss,top_5,val_acc,val_average_class_precision,val_loss,val_top_5
0,0.1906861,0.30422875,3.319986669266912,0.46393448,0.23900227,0.5267521,3.1301780904548755,0.5420225
1,0.32192737,0.5797942,2.7438840397687376,0.6424581,0.33106577,0.60462666,2.7093573307645493,0.6458333
2,0.38716248,0.6410014,2.473531520177976,0.70455074,0.3755858,0.6660643,2.5333779520458646,0.6944337
3,0.42744705,0.67842704,2.3030546094452204,0.74274033,0.38163266,0.68472356,2.5217218554538228,0.70220846
4,0.46841538,0.69185364,2.1507653676153784,0.77511054,0.40037793,0.6937373,2.488985004632369,0.714
```

Online label smoothing
```
0,0.18070589,0.42428213,3.496069971076603,0.44826582,0.25049132,0.7974228,3.057657162348429,0.55951303
1,0.31564245,0.8285898,3.046683876430078,0.6289426,0.33000755,0.857497,2.725422456068693,0.64911145
2,0.37606204,0.87911797,2.8607515232079077,0.6928102,0.37437642,0.89467704,2.568773751673491,0.6812888
3,0.41867435,0.925745,2.7439316036092944,0.73165447,0.4048375,0.9460702,2.437526372319834,0.7142749
4,0.45842063,0.9479143,2.6512795045167152,0.7641556,0.40309903,0.94201684,2.4495919761450393,0.70972437
```

File: plugins/crop_generators/sampled_crop_generator.py [hunk 1]
```diff
@@ -115,20 +115,21 @@ def generator(self, sess, global_step=0):
             for i in range(len(batch_paths)):
                 x[i, :, :, :] = images[i]
 
-            yield (x, tf.keras.utils.to_categorical(y, num_classes=self.num_classes))
+            yield x, tf.keras.utils.to_categorical(y, num_classes=self.num_classes)
 
         image_loader.close()
 
-    def generate(self):
+    def generate(self, sess):
```
Comment (jccaicedo) [inline plugins/crop_generators/sampled_crop_generator.py:122]: Same comment here. Is `sess` necessary? If it is part of the interface for compatibility, that sounds reasonable to me. Just checking.

**Review by jccaicedo — APPROVED** [review]
I'm glad that the change works in practice, that's the most important thing.
It seems like the change was fairly easy, and resulted in less code, which is great!

Some style updates revealed a "hack" that we need to remove asap. There are other couple of questions related to the `sess` parameter, but that seems minor to me. Please, address these issues and feel free to merge!

File: plugins/crop_generators/sampled_crop_generator.py (hunk 1 above)
Comment (Arkkienkeli) [reply plugins/crop_generators/sampled_crop_generator.py:122]: It is a part of interface compatibility.

File: plugins/crop_generators/full_image_crop_generator.py [hunk 2]
```diff
@@ -107,26 +105,26 @@ def generator(self, sess, global_step=0):
                 x[i, :, :, :] = images[i]
 
             inputs = [x, np.asarray([[0,0,1,1]]*len(y)), np.arange(0, len(y))]
-            yield (inputs, tf.keras.utils.to_categorical(y, num_classes=self.num_classes))
+            yield inputs, tf.keras.utils.to_categorical(y, num_classes=self.num_classes)
 
         image_loader.close()
 
-    def generate(self):
+    def generate(self, sess):
```
[the remark this answers is not in the document]
Comment (Arkkienkeli) [reply plugins/crop_generators/full_image_crop_generator.py:112]: It is a part of interface compatibility.

File: plugins/crop_generators/individual_channel_cropgen.py [hunk 3]
```diff
@@ -37,7 +37,8 @@ def __init__(self, config, dset, mode="Training"):
         # Load metadata
         self.all_cells = pd.read_csv(os.path.join(self.directory, "expanded_sc_metadata_tengenes.csv"))
         # ALPHA SET HACK:
```
[the remark this answers is not in the document]
Comment (Arkkienkeli) [reply plugins/crop_generators/individual_channel_cropgen.py:39]: I guess it could be a separate issue.
